File: c/trees/bst_test.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "bst_test.h"
#include "bst.h"
#include "traversal.h"
#include "../test/assert.h"

void BinarySearchTree_test(int *passed, int *total) {
    printf("\n\nBinary Search Tree\n\n");

    // create the tree to test
    BinarySearchTree *root = malloc(sizeof(BinarySearchTree));
    BinarySearchTree_init(root, NULL, 50);
    BinarySearchTree_insert(root, 10);
    BinarySearchTree_insert(root, 30);
    BinarySearchTree_insert(root, 42);
    BinarySearchTree_insert(root, 15);
    BinarySearchTree_insert(root, 3);
    BinarySearchTree_insert(root, 60);
    BinarySearchTree_insert(root, 80);
    BinarySearchTree_insert(root, 83);
    BinarySearchTree_insert(root, 24);
    BinarySearchTree_insert(root, 73);
    BinarySearchTree_insert(root, 87);
    BinarySearchTree_insert(root, 100);

    // verify max and min of original tree, contains
    *passed += ASSERT_INT_EQUALS(100, BinarySearchTree_max(root), "max");
    *passed += ASSERT_INT_EQUALS(3, BinarySearchTree_min(root), "min");
    *passed += ASSERT_TRUE(BinarySearchTree_contains(root, 15), "contains positive");
    *passed += ASSERT_FALSE(BinarySearchTree_contains(root, 20), "contains negative");
    *total += 4;

    // initial insertions
    char *result = malloc(sizeof(char));

    Traversal_inorder_bst(result, root);

    *passed += ASSERT_STR_EQUALS("3 10 15 24 30 42 50 60 73 80 83 87 100", result, "insertion");
    *total += 1;
    memset(result, 0, strlen(result));

    // leaf removal
    BinarySearchTree_remove(root, 3);

    Traversal_inorder_bst(result, root);

    *passed += ASSERT_STR_EQUALS("10 15 24 30 42 50 60 73 80 83 87 100", result, "leaf removal");
    *total += 1;
    memset(result, 0, strlen(result));

    // node with one child removal
    BinarySearchTree_remove(root, 87);

    Traversal_inorder_bst(result, root);

    *passed += ASSERT_STR_EQUALS("10 15 24 30 42 50 60 73 80 83 100", result, "single child removal");
    *total += 1;
    memset(result, 0, strlen(result));

    // node with two children removal
    BinarySearchTree_remove(root, 30);

    Traversal_inorder_bst(result, root);

    *passed += ASSERT_STR_EQUALS("10 15 24 42 50 60 73 80 83 100", result, "two children removal");
    *total += 1;
    memset(result, 0, strlen(result));

    // root removal
    BinarySearchTree_remove(root, 50);

    Traversal_inorder_bst(result, root);

    *passed += ASSERT_STR_EQUALS("10 15 24 42 60 73 80 83 100", result, "root removal");
    *total += 1;
    memset(result, 0, strlen(result));

    free(result);
    free(root);
}

File: haskell/test/Tree/BSTTest.hs
module Tree.BSTTest where

import Test.Framework
import Test.Framework.Providers.HUnit
import Test.HUnit ((@?=))

import Tree.BST as BST
import Tree.Traversal as Traversal

bstTest :: Test
bstTest = let -- this will be better if I convert the project to be monadic
            t1 = BST.singleton 50
            t2 = BST.insert t1 10
            t3 = BST.insert t2 30
            t4 = BST.insert t3 42
            t5 = BST.insert t4 15
            t6 = BST.insert t5 3
            t7 = BST.insert t6 60
            t8 = BST.insert t7 80
            t9 = BST.insert t8 83
            t10 = BST.insert t9 24
            t11 = BST.insert t10 73
            t12 = BST.insert t11 87
            t = BST.insert t12 100 -- finalized initial tree
          in
          testGroup "BST"
                    [ testCase "Insert" $ Traversal.inorder t @?= [3, 10, 15, 24, 30, 42, 50, 60, 73, 80, 83, 87, 100]
                    , testCase "Remove leaf" $ Traversal.inorder $ delete t 3 @?= [10, 15, 24, 30, 42, 50, 60, 73, 80, 83, 87, 100]
                    , testCase "Remove node with single child" $ Traversal.inorder $ delete t 10 @?= [3, 15, 24, 30, 42, 50, 60, 73, 80, 83, 87, 100]
                    , testCase "Remove node with two children" $ Traversal.inorder $ delete t 30 @?= [3, 10, 15, 24, 42, 50, 60, 73, 80, 83, 87, 100]
                    , testCase "Remove root" $ Traversal.inorder $ delete t 50 @?= [3, 10, 15, 24, 30, 42, 60, 73, 80, 83, 87, 100]
                    , testCase "Contains positive" $ BST.contains t 24 @?= True
                    , testCase "Contains negative" $ BST.contains t 25 @?= False
                    ]
